File: src/org/cappuccino_project/ide/intellij/plugin/indices/ObjJIndexService.java
package org.cappuccino_project.ide.intellij.plugin.indices;

import com.intellij.openapi.util.Pair;
import com.intellij.psi.stubs.IndexSink;

import org.cappuccino_project.ide.intellij.plugin.psi.ObjJBlock;
import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJClassType;
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodCallPsiUtil;
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJProtocolDeclarationPsiUtil;
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJTreeUtil;
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.*;
import org.cappuccino_project.ide.intellij.plugin.utils.ArrayUtils;
import org.cappuccino_project.ide.intellij.plugin.utils.ObjJFileUtil;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import static org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils.EMPTY_SELECTOR;
import static org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils.SELECTOR_SYMBOL;

public class ObjJIndexService extends StubIndexService {

    private static final Logger LOGGER = Logger.getLogger("#objj.ObjJIndexService");

    ObjJIndexService() {
     //   Logger.getGlobal().log(Level.INFO, "Creating ObjJIndexService");
    }


    /**
     * Index method header stubs
     * @param methodHeaderStub method header stub to index
     * @param indexSink index sink
     */
    @Override
    public void indexMethod(ObjJMethodHeaderStub methodHeaderStub, IndexSink indexSink) {
        final String selector = methodHeaderStub.getSelectorString();
        if (selector.equals(EMPTY_SELECTOR) || selector.equals(EMPTY_SELECTOR + SELECTOR_SYMBOL)) {
            LOGGER.log(Level.SEVERE, "Method header has no selector to index");
            return;
        }
        if (selector.isEmpty()) {
            LOGGER.log(Level.SEVERE, "Method stub returned with an empty selector");
            return;
        }


        try {
            LOGGER.log(Level.INFO,"Indexing Method - ["+methodHeaderStub.getContainingClassName()+" "+selector + "]");
            indexSink.occurrence(ObjJUnifiedMethodIndex.KEY, selector);
            LOGGER.log(Level.INFO,"Indexed Method - ["+methodHeaderStub.getContainingClassName()+" "+selector + "]");
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Failed to index selector with error: "+e.getLocalizedMessage());
        }

        final String className = methodHeaderStub.getContainingClassName();
        if (!ObjJMethodCallPsiUtil.isUniversalMethodCaller(className)) {
            try {
                indexSink.occurrence(ObjJClassMethodIndex.getInstance().getKey(), className);
            } catch (Exception e) {
                LOGGER.log(Level.SEVERE, "Failed to index class&selector tuple with error: "+e.getLocalizedMessage());
            }
        }
    }

    /**
     * Indexes instance variable stubs
     * @param variableDeclarationStub instance variable stub to index
     * @param indexSink index sink
     */
    public void indexInstanceVariable(@NotNull
                                              ObjJInstanceVariableDeclarationStub variableDeclarationStub, IndexSink indexSink) {
        indexSink.occurrence(ObjJInstanceVariablesByClassIndex.getInstance().getKey(), variableDeclarationStub.getContainingClass());
        indexSink.occurrence(ObjJInstanceVariablesByNameIndex.getInstance().getKey(), variableDeclarationStub.getVariableName());
     //   LOGGER.log(Level.INFO, "Indexing instance variable <"+variableDeclarationStub.getVariableName()+"> for class <"+variableDeclarationStub.getContainingClass()+">");
        if (variableDeclarationStub.getGetter() != null && !variableDeclarationStub.getGetter().isEmpty()) {
            indexSink.occurrence(ObjJClassInstanceVariableAccessorMethodIndex.getInstance().getKey(), variableDeclarationStub.getGetter());
        }
        if (variableDeclarationStub.getSetter() != null && !variableDeclarationStub.getSetter().isEmpty()) {
            indexSink.occurrence(ObjJClassInstanceVariableAccessorMethodIndex.getInstance().getKey(), variableDeclarationStub.getSetter());
        }
    }

    /**
     * Indexes virtual methods from accessor methods
     * @param property accessor property potentially containing virtual methods
     * @param indexSink index sink
     */
    @Override
    public void indexAccessorProperty(@NotNull
                                              ObjJAccessorPropertyStub property, IndexSink indexSink) {
        LOGGER.log(Level.INFO, "Indexing Accessor Property for var <"+property.getContainingClassName()+":"+property.getVariableName()+">");
        if (property.getGetter() != null) {
            LOGGER.log(Level.INFO, "ObjJIndex: Indexing accessorProperty ["+property.getContainingClass()+" "+ property.getGetter() + "]");
            indexSink.occurrence(ObjJUnifiedMethodIndex.KEY, property.getGetter());
        }
        if (property.getSetter() != null) {
            LOGGER.log(Level.INFO, "ObjJIndex: Indexing accessorProperty ["+property.getContainingClass()+" "+ property.getSetter() + "]");
            indexSink.occurrence(ObjJUnifiedMethodIndex.KEY, property.getSetter());
        }
    }

    /**
     * Indexes class declarations
     * @param stub class declaration stub
     * @param indexSink index sink
     */
    @Override
    public void indexClassDeclaration(@NotNull
                                              ObjJClassDeclarationStub stub, @NotNull IndexSink indexSink) {
        if (stub.getClassName().isEmpty()) {
         //   LOGGER.log(Level.INFO, "ClassName is empty in class declaration for index");
            return;
        }
        indexSink.occurrence(ObjJClassDeclarationsIndex.getInstance().getKey(), stub.getClassName());
        if (stub instanceof ObjJImplementationStub) {
            indexImplementationClassDeclaration((ObjJImplementationStub)stub, indexSink);
        } else if (stub instanceof  ObjJProtocolDeclarationStub){
            indexSink.occurrence(ObjJProtocolDeclarationsIndex.getInstance().getKey(), stub.getClassName());
        }
        for (Object protocol : stub.getInheritedProtocols()) {
            if (protocol instanceof String) {
                indexSink.occurrence(ObjJClassInheritanceIndex.getInstance().getKey(), (String)protocol);
            }
        }
    }

    private void indexImplementationClassDeclaration(@NotNull ObjJImplementationStub implementationStub, @NotNull IndexSink indexSink) {
        if (implementationStub.isCategory()) {
            indexSink.occurrence(ObjJClassInheritanceIndex.getInstance().getKey(), implementationStub.getClassName());
            indexSink.occurrence(ObjJImplementationCategoryDeclarationsIndex.getInstance().getKey(),implementationStub.getClassName());
        } else if (implementationStub.getSuperClassName() != null && !implementationStub.getSuperClassName().equals(ObjJClassType.CPOBJECT)) {
         //   LOGGER.log(Level.INFO, "Setting super class to: " + implementationStub.getSuperClassName());
            indexSink.occurrence(ObjJClassInheritanceIndex.getInstance().getKey(), implementationStub.getSuperClassName());
        }
        indexSink.occurrence(ObjJImplementationDeclarationsIndex.getInstance().getKey(), implementationStub.getClassName());
    }

    /**
     * Indexes selector literal method as possible inline method declaration
     * @param selectorLiteral selector literal
     * @param indexSink index sink
     */
    @Override
    public void indexSelectorLiteral(@NotNull
                                              ObjJSelectorLiteralStub selectorLiteral, @NotNull IndexSink indexSink) {
        indexSink.occurrence(ObjJSelectorInferredMethodIndex.getInstance().getKey(), selectorLiteral.getSelectorString());
    }

    @Override
    public void indexFunctionDeclaration(@NotNull
                                                 ObjJFunctionDeclarationElementStub functionDeclaration, IndexSink indexSink) {
        if (functionDeclaration.getFunctionName() == null) {
            //noinspection unchecked
            LOGGER.log(Level.INFO, "function("+ ArrayUtils.join(functionDeclaration.getParamNames()) + ") has no function name");
            return;
        }
        LOGGER.log(Level.INFO, "Indexing function: <"+functionDeclaration.getFunctionName()+">");
        indexSink.occurrence(ObjJFunctionsIndex.getInstance().getKey(), functionDeclaration.getFunctionName());
        LOGGER.log(Level.INFO, "Did Index function: <"+functionDeclaration.getFunctionName()+">");
    }

    @Override
    public void indexMethodCall(ObjJMethodCallStub methodHeaderStub, IndexSink indexSink) {
        LOGGER.log(Level.INFO, "Indexing method call  <["+methodHeaderStub.getCallTarget()+" "+methodHeaderStub.getSelectorString()+"]>");
        indexSink.occurrence(ObjJMethodCallIndex.getInstance().getKey(), methodHeaderStub.getSelectorString());
        LOGGER.log(Level.INFO, "Indexed method call.");
    }


    /**
     * Index global declaration
     * @param globalVariableDeclaration selector literal
     * @param indexSink index sink
     */
    @Override
    public void indexGlobalVariableDeclaration(@NotNull
                                                       ObjJGlobalVariableDeclarationStub globalVariableDeclaration, @NotNull IndexSink indexSink) {
        if (globalVariableDeclaration.getFileName() != null) {
            indexSink.occurrence(ObjJGlobalVariablesByFileNameIndex.getInstance().getKey(), globalVariableDeclaration.getFileName());
        }
        indexSink.occurrence(ObjJGlobalVariableNamesIndex.getInstance().getKey(), globalVariableDeclaration.getVariableName());
    }

    public void indexImport(@NotNull ObjJImportStub stub, IndexSink indexSink) {

    }

    public void indexFile(@NotNull ObjJFileStub fileStub, @NotNull IndexSink indexSink) {
        LOGGER.log(Level.INFO, "Indexing file by name: "+fileStub.getFileName());
        indexSink.occurrence(ObjJFilesByNameIndex.getInstance().getKey(), fileStub.getFileName());
    }

    public void indexVarTypeId(@NotNull ObjJVarTypeIdStub stub, IndexSink indexSink) {

    }

    public void indexVariableName(@NotNull ObjJVariableNameStub stub, @NotNull IndexSink indexSink) {
        final String containingFileName = ObjJFileUtil.getContainingFileName(stub.getPsi().getContainingFile());
        if (containingFileName == null) {
            LOGGER.log(Level.SEVERE, "Cannot index variable name, containing file name is null");
            return;
        } else {
            LOGGER.log(Level.INFO, "Indexing variable: "+containingFileName+": "+stub.getVariableName());
        }
        indexSink.occurrence(ObjJVariableNameByScopeIndex.KEY, containingFileName+"-ALL");
        List<Pair<Integer,Integer>> blockRanges = stub.getContainingBlockRanges();
        if (blockRanges.isEmpty()) {
            indexSink.occurrence(ObjJVariableNameByScopeIndex.KEY, containingFileName+"-TOP");
        }

        for (Pair<Integer,Integer> blockRange : blockRanges) {
            indexSink.occurrence(ObjJVariableNameByScopeIndex.KEY, ObjJVariableNameByScopeIndex.getIndexKey(containingFileName, blockRange));
        }
    }

}

File: src/org/cappuccino_project/ide/intellij/plugin/indices/ObjJIndexService.kt
package org.cappuccino_project.ide.intellij.plugin.indices

import com.intellij.psi.stubs.IndexSink
import org.cappuccino_project.ide.intellij.plugin.lang.ObjJFile
import org.cappuccino_project.ide.intellij.plugin.psi.*
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJClassDeclarationElement
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJFunctionDeclarationElement
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJMethodHeaderDeclaration

import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJClassType
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.*
import org.cappuccino_project.ide.intellij.plugin.utils.ObjJFileUtil

import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils.EMPTY_SELECTOR
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils.SELECTOR_SYMBOL
import org.cappuccino_project.ide.intellij.plugin.psi.utils.isUniversalMethodCaller

class ObjJIndexService
//private static final Logger LOGGER = Logger.getLogger("#objj.ObjJIndexService");

internal constructor()//   Logger.getGlobal().log(Level.INFO, "Creating ObjJIndexService");
    : StubIndexService() {

    /**
     * Index method header stubs
     * @param methodHeaderStub method header stub to index
     * @param indexSink index sink
     */
    override fun indexMethod(methodHeaderStub: ObjJMethodHeaderStub, indexSink: IndexSink) {
        val selector = methodHeaderStub.selectorString
        if (selector == EMPTY_SELECTOR || selector == EMPTY_SELECTOR + SELECTOR_SYMBOL) {
            //LOGGER.log(Level.SEVERE, "Method header has no selector to index");
            return
        }
        if (selector.isEmpty()) {
            //LOGGER.log(Level.SEVERE, "Method stub returned with an empty selector");
            return
        }


        try {
            //LOGGER.log(Level.INFO,"Indexing Method - ["+methodHeaderStub.getContainingClassName()+" "+selector + "]");
            indexSink.occurrence<ObjJMethodHeaderDeclaration<*>, String>(ObjJUnifiedMethodIndex.KEY, selector)
            //LOGGER.log(Level.INFO,"Indexed Method - ["+methodHeaderStub.getContainingClassName()+" "+selector + "]");
        } catch (e: Exception) {
            //LOGGER.log(Level.SEVERE, "Failed to index selector with error: "+e.getLocalizedMessage());
        }

        val selectorBuilder = StringBuilder()
        for (subSelector in methodHeaderStub.selectorStrings) {
            selectorBuilder.append(subSelector).append(ObjJMethodPsiUtils.SELECTOR_SYMBOL)
            indexSink.occurrence<ObjJMethodHeaderDeclaration<*>, String>(ObjJMethodFragmentIndex.KEY, selectorBuilder.toString())
        }

        val className = methodHeaderStub.containingClassName
        if (!isUniversalMethodCaller(className)) {
            try {
                indexSink.occurrence<ObjJMethodHeader, String>(ObjJClassMethodIndex.instance.getKey(), className)
            } catch (e: Exception) {
                //LOGGER.log(Level.SEVERE, "Failed to index class&selector tuple with error: "+e.getLocalizedMessage());
            }

        }
    }

    /**
     * Indexes instance variable stubs
     * @param variableDeclarationStub instance variable stub to index
     * @param indexSink index sink
     */
    override fun indexInstanceVariable(variableDeclarationStub: ObjJInstanceVariableDeclarationStub, indexSink: IndexSink) {
        indexSink.occurrence<ObjJInstanceVariableDeclaration, String>(ObjJInstanceVariablesByClassIndex.instance.getKey(), variableDeclarationStub.containingClass)
        indexSink.occurrence<ObjJInstanceVariableDeclaration, String>(ObjJInstanceVariablesByNameIndex.instance.getKey(), variableDeclarationStub.variableName)
        //   //LOGGER.log(Level.INFO, "Indexing instance variable <"+variableDeclarationStub.getVariableName()+"> for class <"+variableDeclarationStub.getContainingClass()+">");
        if (variableDeclarationStub.getter != null && !variableDeclarationStub.getter!!.isEmpty()) {
            indexSink.occurrence<ObjJInstanceVariableDeclaration, String>(ObjJClassInstanceVariableAccessorMethodIndex.instance.getKey(), variableDeclarationStub.getter!!)
        }
        if (variableDeclarationStub.setter != null && !variableDeclarationStub.setter!!.isEmpty()) {
            indexSink.occurrence<ObjJInstanceVariableDeclaration, String>(ObjJClassInstanceVariableAccessorMethodIndex.instance.getKey(), variableDeclarationStub.setter!!)
        }
    }

    /**
     * Indexes virtual methods from accessor methods
     * @param property accessor property potentially containing virtual methods
     * @param indexSink index sink
     */
    override fun indexAccessorProperty(property: ObjJAccessorPropertyStub, indexSink: IndexSink) {
        //LOGGER.log(Level.INFO, "Indexing Accessor Property for var <"+property.getContainingClassName()+":"+property.getVariableName()+">");
        if (property.getter != null) {
            //LOGGER.log(Level.INFO, "ObjJIndex: Indexing accessorProperty ["+property.getContainingClass()+" "+ property.getGetter() + "]");
            indexSink.occurrence<ObjJMethodHeaderDeclaration<*>, String>(ObjJUnifiedMethodIndex.KEY, property.getter!!)
        }
        if (property.setter != null) {
            //LOGGER.log(Level.INFO, "ObjJIndex: Indexing accessorProperty ["+property.getContainingClass()+" "+ property.getSetter() + "]");
            indexSink.occurrence<ObjJMethodHeaderDeclaration<*>, String>(ObjJUnifiedMethodIndex.KEY, property.setter!!)
        }
    }

    /**
     * Indexes class declarations
     * @param stub class declaration stub
     * @param indexSink index sink
     */
    override fun indexClassDeclaration(stub: ObjJClassDeclarationStub<*>, indexSink: IndexSink) {
        if (stub.className.isEmpty()) {
            //   //LOGGER.log(Level.INFO, "ClassName is empty in class declaration for index");
            return
        }
        indexSink.occurrence<ObjJClassDeclarationElement<*>, String>(ObjJClassDeclarationsIndex.instance.getKey(), stub.className)
        if (stub is ObjJImplementationStub) {
            indexImplementationClassDeclaration(stub, indexSink)
        } else if (stub is ObjJProtocolDeclarationStub) {
            indexSink.occurrence<ObjJProtocolDeclaration, String>(ObjJProtocolDeclarationsIndex.instance.getKey(), stub.className)
        }
        for (protocol in stub.inheritedProtocols) {
            indexSink.occurrence<ObjJClassDeclarationElement<*>, String>(ObjJClassInheritanceIndex.instance.getKey(), protocol)
        }
    }

    private fun indexImplementationClassDeclaration(implementationStub: ObjJImplementationStub, indexSink: IndexSink) {
        if (implementationStub.isCategory) {
            indexSink.occurrence<ObjJClassDeclarationElement<*>, String>(ObjJClassInheritanceIndex.instance.getKey(), implementationStub.className)
            indexSink.occurrence<ObjJImplementationDeclaration, String>(ObjJImplementationCategoryDeclarationsIndex.instance.getKey(), implementationStub.className)
        } else if (implementationStub.superClassName != null && implementationStub.superClassName != ObjJClassType.CPOBJECT) {
            //   //LOGGER.log(Level.INFO, "Setting super class to: " + implementationStub.getSuperClassName());
            indexSink.occurrence<ObjJClassDeclarationElement<*>, String>(ObjJClassInheritanceIndex.instance.getKey(), implementationStub.superClassName!!)
        }
        indexSink.occurrence<ObjJImplementationDeclaration, String>(ObjJImplementationDeclarationsIndex.instance.getKey(), implementationStub.className)
    }

    /**
     * Indexes selector literal method as possible inline method declaration
     * @param selectorLiteral selector literal
     * @param indexSink index sink
     */
    override fun indexSelectorLiteral(selectorLiteral: ObjJSelectorLiteralStub, indexSink: IndexSink) {
        indexSink.occurrence<ObjJSelectorLiteral, String>(ObjJSelectorInferredMethodIndex.instance.getKey(), selectorLiteral.selectorString)
        val stringBuilder = StringBuilder()
        for (selector in selectorLiteral.selectorStrings) {
            stringBuilder.append(selector).append(ObjJMethodPsiUtils.SELECTOR_SYMBOL)
            indexSink.occurrence<ObjJMethodHeaderDeclaration<*>, String>(ObjJMethodFragmentIndex.KEY, stringBuilder.toString())
        }
    }

    override fun indexFunctionDeclaration(variableDeclarationStub: ObjJFunctionDeclarationElementStub<*>, indexSink: IndexSink) {
        //LOGGER.log(Level.INFO, "Indexing function: <"+functionDeclaration.getFunctionName()+">");
        indexSink.occurrence<ObjJFunctionDeclarationElement<*>, String>(ObjJFunctionsIndex.instance.getKey(), variableDeclarationStub.functionName)
        //LOGGER.log(Level.INFO, "Did Index function: <"+functionDeclaration.getFunctionName()+">");
    }

    override fun indexMethodCall(methodHeaderStub: ObjJMethodCallStub, indexSink: IndexSink) {
        //LOGGER.log(Level.INFO, "Indexing method call  <["+methodHeaderStub.getCallTarget()+" "+methodHeaderStub.getSelectorString()+"]>");
        indexSink.occurrence<ObjJMethodCall, String>(ObjJMethodCallIndex.instance.getKey(), methodHeaderStub.selectorString)
        //LOGGER.log(Level.INFO, "Indexed method call.");
    }


    /**
     * Index global declaration
     * @param globalVariableDeclaration selector literal
     * @param indexSink index sink
     */
    override fun indexGlobalVariableDeclaration(globalVariableDeclaration: ObjJGlobalVariableDeclarationStub, indexSink: IndexSink) {
        if (globalVariableDeclaration.fileName != null) {
            indexSink.occurrence<ObjJGlobalVariableDeclaration, String>(ObjJGlobalVariablesByFileNameIndex.instance.getKey(), globalVariableDeclaration.fileName!!)
        }
        indexSink.occurrence<ObjJGlobalVariableDeclaration, String>(ObjJGlobalVariableNamesIndex.instance.getKey(), globalVariableDeclaration.variableName)
    }

    override fun indexImport(stub: ObjJImportStub<*>, indexSink: IndexSink) {

    }

    override fun indexFile(stub: ObjJFileStub, sink: IndexSink) {
        //LOGGER.log(Level.INFO, "Indexing file by name: "+fileStub.getFileName());
        sink.occurrence<ObjJFile, String>(ObjJFilesByNameIndex.instance.getKey(), stub.fileName)
    }

    override fun indexVarTypeId(stub: ObjJVarTypeIdStub, indexSink: IndexSink) {

    }

    override fun indexVariableName(stub: ObjJVariableNameStub, indexSink: IndexSink) {
        val containingFileName = ObjJFileUtil.getContainingFileName(stub.psi.containingFile)
                ?: //LOGGER.log(Level.SEVERE, "Cannot index variable name, containing file name is null");
                return
// else {
        //LOGGER.log(Level.INFO, "Indexing variable: "+containingFileName+": "+stub.getVariableName());
        //}
        indexSink.occurrence<ObjJVariableName, String>(ObjJVariableNameByScopeIndex.KEY, "$containingFileName-ALL")
        val blockRanges = stub.containingBlockRanges
        if (blockRanges.isEmpty()) {
            indexSink.occurrence<ObjJVariableName, String>(ObjJVariableNameByScopeIndex.KEY, "$containingFileName-TOP")
        }

        for (blockRange in blockRanges) {
            indexSink.occurrence<ObjJVariableName, String>(ObjJVariableNameByScopeIndex.KEY, ObjJVariableNameByScopeIndex.getIndexKey(containingFileName, blockRange))
        }
    }

    companion object {
        const val INDEX_VERSION = 1
    }
}
